### ISSUE TITLE
Deployment not working

### ISSUE THREAD

**Issue by vsmart** [rubycorns/rorganize.it#549]
Pushing to production master does not seem to kick off deployment at the moment.

**Comment by lislis** [comment]
I ssh'ed into uberspace and the git log seems fine and up to date but it's not showing the most up to date commit on the live site. But I didn't look further into it.

**Comment by til** [comment]
It's a bit hidden, but there is a log file on the server that contains
some info what happens during a deploy, in log/deploy.log. Here are the
last lines from it, apparently a bundle install problem, does that ring
a bell with anyone? Maybe upgrading json or related gems could fix it?


==== Tue May 30 20:20:04 CEST 2017 ====
files changed: 7
The git source `git://github.com/stefanpenner/country_select.git` uses the `git` protocol, which transmits data without encryption. Disable this warning with `bundle config git.allow_insecure true`, or switch to the `https` protocol to keep your data secure.
Fetching gem metadata from https://rubygems.org/............
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Using rake 11.1.1
Using i18n 0.7.0
Installing json 1.8.3 with native extensions
Using minitest 5.8.4
Using thread_safe 0.3.5
Using builder 3.2.2
Using erubis 2.7.0
Using mini_portile2 2.0.0
Using rack 1.6.4
Using mime-types 2.99.1
Using arel 6.0.3
Using addressable 2.3.6
Using bcrypt 3.1.11
Using chunky_png 1.3.1
Using fssm 0.2.10
Using sass 3.2.19
Using currencies 0.4.2
Using unicode_utils 1.4.0
Using daemons 1.1.9
Using orm_adapter 0.5.0
Using thor 0.19.1
Using eventmachine 1.0.8
Using execjs 2.6.0
Using multipart-post 2.0.0
Using geocoder 1.2.14
Using tilt 1.4.1
Using hashie 3.3.2
Using hike 1.2.3
Using multi_json 1.11.2
Using jwt 1.2.0
Using leaflet-rails 0.7.7
Using redcarpet 3.2.3
Using subexec 0.2.3
Using modernizr-rails 2.6.2.3
Using multi_xml 0.5.5
Using pg 0.17.1
Using bundler 1.14.6
Using will_paginate 3.0.5
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

current directory:
/home/ror/rorganize.it/vendor/bundle/ruby/2.4.0/gems/json-1.8.3/ext/json/ext/generator
/package/host/localhost/ruby-2.4.1/bin/ruby -r
./siteconf20170530-4464-1cjn34d.rb extconf.rb
creating Makefile

current directory:
/home/ror/rorganize.it/vendor/bundle/ruby/2.4.0/gems/json-1.8.3/ext/json/ext/generator
make "DESTDIR=" clean

current directory:
/home/ror/rorganize.it/vendor/bundle/ruby/2.4.0/gems/json-1.8.3/ext/json/ext/generator
make "DESTDIR="
compiling generator.c
generator.c: In function ‘generate_json’:
generator.c:861: error: ‘rb_cFixnum’ undeclared (first use in this function)
generator.c:861: error: (Each undeclared identifier is reported only once
generator.c:861: error: for each function it appears in.)
generator.c:863: error: ‘rb_cBignum’ undeclared (first use in this function)
At top level:
cc1: warning: unrecognized command line option "-Wno-self-assign"
cc1: warning: unrecognized command line option "-Wno-constant-logical-operand"
cc1: warning: unrecognized command line option "-Wno-parentheses-equality"
cc1: warning: unrecognized command line option "-Wno-tautological-compare"
make: *** [generator.o] Error 1

make failed, exit code 2

Gem files will remain installed in
/home/ror/rorganize.it/vendor/bundle/ruby/2.4.0/gems/json-1.8.3 for inspection.
Results logged to
/home/ror/rorganize.it/vendor/bundle/ruby/2.4.0/extensions/x86_64-linux/2.4.0-static/json-1.8.3/gem_make.out

An error occurred while installing json (1.8.3), and Bundler cannot continue.
Make sure that `gem install json -v '1.8.3'` succeeds before bundling.

**Comment by vsmart** [comment]
I saw this json error recently when I tried to use the app with ruby 2.4 -- apparently 2.4 does not work with json gem. https://github.com/rails/rails/issues/27450
But why would the Ruby version be updated to 2.4? :question:

**Comment by til** [comment]
Aha, good to know. It seems the default version on the server is now 2.4

[ror@schedar ~]$ ruby -v
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]

According to https://wiki.uberspace.de/development:ruby?s[]=ruby we
could theoretically set it explicitly to an older version, but I think
it'd be better to move forward and fix our app to work with ruby 2.4. Do
you have an idea what could fix it? Upgrade whatever gem the json gem
depends on maybe, or just upgrade the json gem?

**Comment by vsmart** [comment]
I think we need to update Rails to 4.2.8, and then use Ruby 2.4. I will make a PR!

**Comment by lislis** [comment]
Reopening
After deploying the server is now throwing a 503.
I stopped daemontools and ran `~/service/autostart/run` manually and got
```
Your bundle is locked to rake (12.0.0), but that version could not be found in any of the sources listed in your Gemfile. If you haven't changed sources, that means the author of rake (12.0.0) has removed it. You'll need to update your bundle to a different version of rake (12.0.0) that hasn't been removed in order to install.
```
The `bundle` runs through without any problems though but this error persists and keeps the server from running.
Any ideas?

**Comment by til** [comment]
I think that error normally occurs when not starting rake from the
bundle, ie. calling just `rake` instead of `bundle exec rake` or
`bin/rake`.

... ok this time the reason is a different one. Following the
breadcrumbs from daemontools to see what is executed, this comes up:


[ror@schedar ~]$ cat ~/rorganize.it/config/autostart
#!/bin/sh
cd ~/rorganize.it/
export RAILS_ENV=production
export PGHOST=/home/ror/tmp
echo "autostarting thin app server"
/package/host/localhost/ruby-2.3.1/bin/bundle exec thin -a 127.0.0.1 -p 34567 start


So the ruby version needs to changed there.

**Comment by til** [comment]
Iirc that file is put there without version control. Do you just want to
edit it?

This would be the path to the current ruby:

[ror@schedar ~]$ which ruby
/package/host/localhost/ruby-2.4.1/bin/ruby


This setup is not perfect, it will fail with the next ruby upgrade again
:(

Probably when executed by daemontools, the PATH is not set, therefore we
cannot just call `ruby` instead of specifying the full path.

**Comment by lislis** [comment]
I was just about about to ask if I should edit it on the server.

I'll open a new issue to document this for the next time

**Comment by til** [comment]
Cool, thanks for fixing this!

ror is up again, yeah. And on the upside, we know the alert system is
working ;)
